waterbox/opera/bizhawk.cpp

```diff
@@ -0,0 +1,402 @@
+{ 
+	_gameFilePath = gameFilePath;
+	_biosFilePath = biosFilePath;
+	_fontFilePath = fontFilePath;
+	_port1Type = port1Type;
+	_port2Type = port2Type;
+ _region = region;
+
+	opera_lr_callbacks_set_environment(retro_environment_callback);
+	opera_lr_callbacks_set_input_state(retro_input_state_callback);
+	opera_lr_callbacks_set_input_poll(retro_input_poll_callback);
+	opera_lr_callbacks_set_audio_sample_batch(retro_audio_sample_batch_callback);
+	opera_lr_callbacks_set_video_refresh(retro_video_refresh_callback);
+
+	retro_set_controller_port_device(0, port1Type);
+	retro_set_controller_port_device(1, port2Type);
+	retro_init();
+
+ // Setting cd callbacks
+	opera_cdrom_set_callbacks(cd_get_size, cd_set_sector, cd_read_sector);
+
+	// Loading game file
+	struct retro_game_info game;
+	game.path = _gameFilePath.c_str();
+	auto loadResult = retro_load_game(&game);
+	if (loadResult == false) { fprintf(stderr, "Could not load game: '%s'\n", _gameFilePath.c_str()); return false; }
+
+	//// Getting memory areas
+	/** 2 = wram, 3 = vram*/
+
+	// WRAM
+	_memoryAreas.wram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_SYSTEM_RAM);
+	_memorySizes.wram = retro_get_memory_size(RETRO_MEMORY_SYSTEM_RAM);
+
+	// VRAM
+	_memoryAreas.vram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_VIDEO_RAM);
+	_memorySizes.vram = retro_get_memory_size(RETRO_MEMORY_VIDEO_RAM);
+
+	// Getting av info
+	struct retro_system_av_info info;
+	retro_get_system_av_info(&info);
+	printf("3DO Framerate: %f\n", info.timing.fps);
+
+	return true;
+}
+
+ECL_EXPORT void opera_get_video(int& w, int& h, int& pitch, uint8_t*& buffer)
+{
+	buffer = (uint8_t*)_videoBuffer;
+	w = _videoWidth;
+	h = _videoHeight;
+	pitch = _videoPitch;
+}
+
+ECL_EXPORT void FrameAdvance(MyFrameInfo* f)
+{
+	 // Setting inputs
+		_port1Value = f->port1;
+		_port2Value = f->port2;
+
+  //printf("Mouse X%d(%d), Y%d(%d), L%d, M%d, B%d\n", _port1Value.mouse.posX, _port1Value.mouse.dX, _port1Value.mouse.posY, _port1Value.mouse.dY, _port1Value.mouse.leftButton, _port1Value.mouse.middleButton, _port1Value.mouse.rightButton);
+		//fflush(stdout);
+
+  // Checking for changes in NVRAM
+		_nvramChanged = false;
+
+  // Checking for lag frames
+		_isLagFrame = true;
+
+		// If resetting, do it now. Otherwise, running a single frame
+		if (f->isReset == 1) retro_reset();
+		else retro_run();
+
+		// Setting if lag frame
+		f->base.Lagged = _isLagFrame ? 1 : 0;
+
+		// Setting video buffer
+		f->base.Width = _videoWidth;
+		f->base.Height = _videoHeight;
+		memcpy(f->base.VideoBuffer, _videoBuffer, sizeof(uint32_t) * _videoWidth * _videoHeight);
+
+		// Setting audio buffer
+		f->base.Samples = _audioSamples;
+		memcpy(f->base.SoundBuffer, _audioBuffer, _audioSamples * sizeof(int16_t) * _CHANNEL_COUNT);
+}
+
+ECL_EXPORT void GetMemoryAreas(MemoryArea *m)
+{
+	int memAreaIdx = 0;
+
+	m[memAreaIdx].Data  = _memoryAreas.wram;
+	m[memAreaIdx].Name  = "Work RAM";

```

Comment:
Isn't this internally `DRAM`? Does it make sense to call this wram here instead of dram?

Reply:
fixed. Also exposing non volatile ram (SaveRAM)